### ISSUE TITLE
[outline view] support a push model

### ISSUE THREAD

**Issue by connectdotz** [Microsoft/vscode#61104]
Right now outline view only supports a pull model, i.e. invoking `DocumentSymbolProvider.provideDocumentSymbol()` when document content changed. However, for custom document symbol providers that prefer to update symbols with different events, such as `onDidSaveTextDocument` or other external events, there didn't seem to be a way to signal/push symbols to the outline view.

Suggest that `DocumentSymbolProvider` can follow the same push pattern, such as with `TextDocumentContentProvider` or `CodeLensProvider`, which utilizes the event model to signal/push changes  from custom providers.

**Comment by connectdotz** [comment]
it's a tradeoff between performance and accuracy. In our example, we are trying to provid the test names for the unnamed test blocks `<function>` (see #60641), which requires parsing the whole file on each `provideDocumentSymbol` call. During active editing, it can generate a lot of activities negatively impacts the responsiveness. 

It is ligit to ask why is it ok to delay such update until the "save"? For our provider/extension, the rationales include:
1. our extension triggers background unit testing upon document save, it is reasonable for the users to assume the test info, as well as test status, will be updated at such time.
1. The name of the test is mainly information purpose, one can argue that a slight delay is probably ok.

**Comment by jrieken** [comment]
> which requires parsing the whole file on each provideDocumentSymbol call

That's what most of them do and that's ok. We debounce requests on the client so that it doesn't happen on each keystroke. If your computation is expensive make sure to move it into a separate process so that you don't block extensions (see LSP). 

We are not planning to go to the push model here but you prevent updates if you want. Nothing prevents your provider from returning an "old" (cached) result and only compute a fresh result every so often.

**Comment by connectdotz** [comment]
the whole idea about this feature request is to address _"what to do when the extension has the data but vscode is not calling..."_ Without push, we can only submit the updated symbols on the **next** `provideDocumentSymbol` call, which is only triggered when the user edit the document. 

Here is a simple workflow to illustrate the user experience:

1. user edit the document => outline didn't change because we are returning the cached (old) symbols
1. user save the document, document is then updated (by the extension) with test status decorators, new code lense, and we now have the updated document symbols => but outline didn't change because vscode is not calling  the extension.
1. if the user wanted to see outline reflecting the latest document state, i.e. with the new symbols, he/she would have to "fake" an edit in order to trigger vscode issuing `provideDocumentSymbol` again...

IMHO, this is not an acceptable user experience. 

In case I haven't made it clear, I wasn't suggesting to replace the current pull model with push, only to supplement it with an additional push mechanism that already supported by other vscode components. This is to allow extensions to properly contributing to outline view when they have off-cycle updates. Your early comments seem to imply that there is a simple workaround thus the push model is not needed, but I am not sure such workaround actually addressed the issue above...

**Comment by jrieken** [comment]
> In case I haven't made it clear, I wasn't suggesting to replace the current pull model with push, only to supplement it with an additional push mechanism that already supported by other vscode components

I understand that and I agree that this isn't an acceptable user experience. But that's also true for the push model extension you are proposing. As a user I expect the outline view to update as I type, not as the extension is taking it's time to recompute the outline. Computing the outline should be a fast and you should try to make your outline provider fast.

**Comment by connectdotz** [comment]
> Computing the outline should be a fast and you should try to make your outline provider fast.

hmm... if I made it sound like we are asking this feature because our extension is slow, let me correct this misconception here. On the contrary, we are able to handle the pull updates just fine as you would have hoped. But just because we can doesn't mean we should, as a developer, I know I could be using user's resource unnecessarily. And as a user, I have been to the situtations too many times that when multiple processes are going at the same time or when I am using a less powerful host, the normally fast and unnoticeable operations suddently became untolerably slow... then you start to ask yourself, do I really need to do all these operations at this point...?

> As a user I expect the outline view to update as I type, not as the extension is taking it's time to recompute the outline.

I would rephrase the question as such: _As a user, do I need the outline view to update as I type even when I am short on CPU resource and can I live with less frequent update when needed...?_

Every prodiver takes time to run, no matter how fast it is, it can't beat the peformance of not running, i.e. zero. Thanks to this discussion, I come to realize the perfect balance between performance vs. accuracy can not be decided by vscode nor extensions as a one-size-fits-all solution, since it is highly subjective and heavily depending on user's environment, it would only make sense to allow the users to decide for themselves. Like code formating, users should be able to say I want my test outline to be updated as I type or when I save. We are capable of supporting both today, if vscode provides a hook, which needs not be through push model, it can simply invoke `provideDocumentSymbol` when document is saved...

Having said all these, if typescript 3.2 release did address the unnamed function symbols ([#58367](https://github.com/Microsoft/vscode/issues/58367#issuecomment-431225051)), I am perfectly ok to walk away from this request and leave the battle to the next lucky extension... 😏
